openage/convert/texture.py

```diff
@@ -144,69 +143,48 @@ def structs(cls):
         return [struct_definition.StructDefinition(cls)]
 
 
-def merge_frames(frames, max_width=0, max_height=0):
+def merge_frames(frames):
     """
     merge all given frames of this slp to a single image file.
 
     frames = [TextureImage, ...]
 
-    returns = TextureImage, [drawn_frames_meta]
+    returns = TextureImage, (width, height) [drawn_frames_meta]
     """
 
     import numpy
 
-    # TODO: actually optimize free space on the texture.
-    # if you ever wanted to implement a combinatoric optimisation
-    # algorithm, go for it, this function just waits for you.
-    # https://en.wikipedia.org/wiki/Bin_packing_problem
-    #
-    # for now, using max values for solving bin packing problem
-    # after determining the maximum frame width and height,
-    # each frame will be placed in a grid with fields of these sizes.
-    # this wastes storage, but works. If your blood boils now because you
-    # wanna optimize for the best alignment, read the above notice again,
-    # and implement a better version.
-
     if len(frames) == 0:
         raise Exception("cannot create texture with empty input frame list")
 
-    # single-frame texture, no merging needed
-    elif len(frames) == 1:
-        cx, cy = frames[0].hotspot
-        w, h = frames[0].width, frames[0].height
-        return frames[0], (w, h), [subtexture_meta(0, 0, w, h, cx, cy)]
+    packer = BestPacker([BinaryTreePacker(margin=1, aspect_ratio=1),
+                         BinaryTreePacker(margin=1, aspect_ratio=97 / 49),
+                         RowPacker(margin=1)])
+    packer.pack(frames)
 
-    # if not predefined, get maximum frame size by checking all frames
-    if max_width == 0 or max_height == 0:
-        max_width = max([teximg.width for teximg in frames])
-        max_height = max([teximg.height for teximg in frames])
+    width, height = packer.width(), packer.height()
+    assert width <= 4096 and height <= 4096, "Texture size limit exceeded"
```

Comment:
those hardcoded values should be placed somewhere else i think, maybe in the `hardcoded.py` file?